homeassistant/components/google_drive/backup.py

```diff
@@ -0,0 +1,273 @@
+    return [
+        GoogleDriveBackupAgent(hass=hass, config_entry=config_entry)
+        for config_entry in hass.config_entries.async_loaded_entries(DOMAIN)
+    ]
+
+
+@callback
+def async_register_backup_agents_listener(
+    hass: HomeAssistant,
+    *,
+    listener: Callable[[], None],
+    **kwargs: Any,
+) -> Callable[[], None]:
+    """Register a listener to be called when agents are added or removed.
+
+    :return: A function to unregister the listener.
+    """
+    hass.data.setdefault(DATA_BACKUP_AGENT_LISTENERS, []).append(listener)
+
+    @callback
+    def remove_listener() -> None:
+        """Remove the listener."""
+        hass.data[DATA_BACKUP_AGENT_LISTENERS].remove(listener)
+
+    return remove_listener
+
+
+class ChunkAsyncStreamIterator:
+    """Async iterator for chunked streams.
+
+    Based on aiohttp.streams.ChunkTupleAsyncStreamIterator, but yields
+    bytes instead of tuple[bytes, bool].
+    """
+
+    __slots__ = ("_stream",)
+
+    def __init__(self, stream: StreamReader) -> None:
+        """Initialize."""
+        self._stream = stream
+
+    def __aiter__(self) -> Self:
+        """Iterate."""
+        return self
+
+    async def __anext__(self) -> bytes:
+        """Yield next chunk."""
+        rv = await self._stream.readchunk()
+        if rv == (b"", False):
+            raise StopAsyncIteration
+        return rv[0]
+
+
+class GoogleDriveBackupAgent(BackupAgent):
+    """Google Drive backup agent."""
+
+    domain = DOMAIN
+
+    def __init__(
+        self, hass: HomeAssistant, config_entry: GoogleDriveConfigEntry
+    ) -> None:
+        """Initialize the cloud backup sync agent."""
+        super().__init__()
+        self.name = config_entry.title
+        self._hass = hass
+        self._folder_id = config_entry.unique_id
+        self._auth: AsyncConfigEntryAuth = config_entry.runtime_data
+
+    async def async_upload_backup(
+        self,
+        *,
+        open_stream: Callable[[], Coroutine[Any, Any, AsyncIterator[bytes]]],
+        backup: AgentBackup,
+        **kwargs: Any,
+    ) -> None:
+        """Upload a backup.
+
+        :param open_stream: A function returning an async iterator that yields bytes.
+        :param backup: Metadata about the backup that should be uploaded.
+        """
+        headers = await self._async_headers()
+        properties = _convert_agent_backup_to_properties(backup)
+        with MultipartWriter() as mpwriter:
+            mpwriter.append_json(
+                {
+                    "name": f"{backup.name} {backup.date}.tar",
+                    "parents": [self._folder_id],
+                    "properties": properties,
+                }
+            )
+            mpwriter.append(await open_stream())
+            headers.update(
+                {"Content-Type": f"multipart/related; boundary={mpwriter.boundary}"}
+            )
+            try:
+                resp = await async_get_clientsession(self._hass).post(
```

Comment:
Does this handle files larger than 5MB safely? From the docs a multipart upload is good for files less than 5mb in size, but files larger should use a resumable file upload (split into chunks). 

https://developers.google.com/drive/api/guides/manage-uploads#resumable